### PR TITLE
Feat/swaps link buy gas

### PR DESCRIPTION
**Pull request by infiniteflower**
## Explanation

This PR resolves [MMS-525](https://consensyssoftware.atlassian.net/browse/MMS-525)

> Problem

> When a user does not have enough gas to complete a transaction, we show them an error message & they are at a dead end.

> Proposed Solution

> We should offer a link to on-ramps so that they can top-up their balance, allowing them to complete swaps. On mobile, this is a hyperlink at the end of the message, saying “Buy more ETH”. We would update “ETH” to match the native token symbol if the user is not on eth mainnet.

1. Adds a link to buy ETH (or whatever native gas token is needed for the network) when the gas is insufficient to complete the swap.

### Before

Screenshot 2023-07-27 at 3 20 00 PM

### After

Screenshot 2023-07-27 at 3 45 33 PM

## Manual Testing Steps

1. Go to Swaps
2. Select the gas token for the current network as the From token
3. Enter in an amount greater or equal to your current balance for From token
4. Observe the link to "Buy ETH"
5. Click the link
6. A new tab should open showing the Portfolio and its Buy page

Make sure to test both with Smart Transactions On and Off

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Review by dan437 — DISMISSED** [review]
Looks good!

**Comment by darkwing** [comment]
Looks pretty good but could you look at the line-height between the text and the link?  The link looks slightly lower:

SCR-20230728-kdfn

**Comment by infiniteflower** [comment]
@darkwing Fixed the misalignment
Screenshot 2023-07-28 at 2 54 31 PM

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (NidhiKJha) [inline ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:31]: Can we add a `className` here or remove this empty string?

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (NidhiKJha) [inline ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:46]: I think we can avoid this inline style by adding style to the class name. wdyt?

**Review by NidhiKJha — COMMENTED** [review]
UI-wise looks really good. Before I approve, just some nits.

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:31]: Cannot remove the empty string as it's a required prop in the `PropType` and TS is enforcing it. Didn't want to modify the underlying component and make this PR bloated and touching too many thing. I could if you think it's a good idea though.

Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:31]: Or I can add a `className` with just empty CSS.

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:46]: For this one I can make a new `className` with the inline styles

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.stories.tsx
Comment (georgewrmarshall) [inline ui/pages/swaps/prepare-swap-page/not-enough-balance.stories.tsx:3]: Nit: remove commented code

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (georgewrmarshall) [inline ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:5]: `SEVERITIES` has been deprecated in favor of the `Severity` enum we are actively trying to reduce it's use through [good first issue and our open source community](https://github.com/MetaMask/metamask-extension/issues/18714). Let me know if you don't see the strike through in VS Code and we can figure out what the settings are :) 

```suggestion
  Severity,
```

Screenshot 2023-07-31 at 9 25 36 AM

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx]: Resolved in 9778cd2

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx]: Resolved in 9778cd2

File: ui/pages/swaps/prepare-swap-page/review-quote.js
Comment (georgewrmarshall) [inline ui/pages/swaps/prepare-swap-page/review-quote.js:1021]: Would this work?
```
{isShowingWarning && (
          <>
            {viewQuotePriceDifferenceWarning}
            {(showInsufficientWarning || tokenBalanceUnavailable) && (
                <BannerAlert
                  title={t('notEnoughBalance')}
                  severity={Severity.Info}
                  description={actionableBalanceErrorMessage}
                  descriptionProps={{ 'data-testid': 'mm-banner-alert-notification-text' }}
                  actionButtonLabel={needsMoreGas ? t('buyMoreAsset', [nativeCurrencySymbol]) : undefined}
                  actionButtonOnClick={
                    needsMoreGas ? () => openBuyCryptoInPdapp() : undefined
                  }
                  marginTop={2}
                />
            )}
          </>
        )}
```

File: ui/pages/swaps/prepare-swap-page/index.scss
Comment (georgewrmarshall) [inline ui/pages/swaps/prepare-swap-page/index.scss:407]: Not sure we want to use css if no styles are added

**Review by georgewrmarshall — COMMENTED** [review]
Hey @infiniteflower, this is looking like a great start I have a couple questions after further review. Is the intention to use the `NotEnoughBalance` component in other areas of the application? Apologies if this is in the ticket you've attached I don't have access even after logging into Jira with my work credentials. If so I think creating another component makes sense but if not would it be possible to just use the `BannerAlert` instance? If the intention is to re-use this component in other places around the application then maybe it makes sense to create it's own folder and put all of the style there?

File: ui/pages/swaps/prepare-swap-page/index.scss
Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/index.scss:407]: This came out of a previous comment from another reviewer: https://github.com/MetaMask/metamask-extension/pull/20241#discussion_r1279378004

Please advise on how you would like to proceed given the TS typing errors that come up.

File: ui/pages/swaps/prepare-swap-page/review-quote.js
Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/review-quote.js:1021]: My original intention with the `NotEnoughBalance` component was simply to speed up development. I pulled it out so I could make a story for it and adjust the styling more easily, rather than constantly reload the extension and retype in the parameters I needed.

Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/review-quote.js:1021]: Please let me know if you want to keep the component as is, or if you want me to put it back inline with the original file.

File: ui/pages/swaps/prepare-swap-page/index.scss
Comment (NidhiKJha) [reply ui/pages/swaps/prepare-swap-page/index.scss:407]: I think you can remove this css, by empty CSS in that comment I meant no css if style is not required. But we can keep the className in the component. 
```suggestion

```

File: ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx
Comment (NidhiKJha) [inline ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:34]: Thanks for adding it here! Just one small nit to remove the unused CSS with this file and it will be good to go.

Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:34]: @NidhiKJha TS will complain if the `className` is removed.

```
Property 'className' is missing in type '{ children: Element; severity: Severity; title: string; }' but required in type '{ [x: string]: any; children: any; className: any; severity?: string | undefined; }'.ts(2741)
```

Please let me know what you like me to do

Comment (NidhiKJha) [reply ui/pages/swaps/prepare-swap-page/not-enough-balance.tsx:34]: This is exactly what I was referring to. To add a className but no CSS as we don't need styling. Thanks, @infiniteflower

**Review by NidhiKJha — DISMISSED** [review]

File: ui/pages/swaps/prepare-swap-page/review-quote.js
Comment (georgewrmarshall) [reply ui/pages/swaps/prepare-swap-page/review-quote.js:1021]: I understand and completetly agree that storybook speeds up development! One option is that we could create a story for `review-quote.js` and feed in the appropriate mock state using the providers to show the `BannerAlert` in a story. This prevents the need to create another component and allows for the component to be viewed in isolation. If that seems out of scope of this ticket I think the route you are taking is good. I would suggest using the `BannerAlert` props to show the action so it is consisten with other banners. I think we would need to include a period between the button and description if it were all on the same line.

Comment (infiniteflower) [reply ui/pages/swaps/prepare-swap-page/review-quote.js]: Thanks, I've updated the component as you suggested. We won't need the period if we go with your approach of using the `BannerAlert` props, since the button will appear on a separate line.

File: ui/pages/swaps/prepare-swap-page/index.scss
Comment (georgewrmarshall) [inline ui/pages/swaps/prepare-swap-page/index.scss:376]: Shouldn't need these styles now that you are using the actionButton props in BannerAlert

**Review by georgewrmarshall — COMMENTED** [review]
Looks good! One suggestion to remove unused CSS. To me this approach minimizes code while aligning with the design system but you're absolutely right it's hard to work on this in isolation without trawling through the app to find the specific component and state. I tried to get the review-quote.stories.js up and running. I think doing that would provide that ability but I'm not as familiar with swaps. Here is where I got up to. I don't think it's necessary for this PR but it might be something worth looking into to improve the developer experience of engineers working on swaps in future

```
import React from 'react';
import { Provider } from 'react-redux';
import testData from '../../../../.storybook/test-data';
import configureStore from '../../../store/store';
import ReviewQuote from './review-quote';

const store = configureStore({
  ...testData,
  metamask: {
    ...testData.metamask,
    swapsState: {
      swapsFeatureFlags: {
        smartTransactions: {
          mobileActive: true,
          extensionActive: true,
        },
        swapRedesign: {
          mobileActive: true,
          extensionActive: true,
        },
      },
      quotes: {
        TEST_AGG_1: {
          trade: {
            from: '0xe18035bf8712672935fdb4e5e431b1a0183d2dfc',
            value: '0x0',
            gas: '0x61a80', // 4e5
            to: '0x881D40237659C251811CEC9c364ef91dC08D300C',
          },
          sourceAmount: '10000000000000000000', // 10e18
          destinationAmount: '20000000000000000000', // 20e18
          error: null,
          sourceToken: '0x6b175474e89094c44da98b954eedeac495271d0f',
          destinationToken: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
          approvalNeeded: null,
          maxGas: 600000,
          averageGas: 120000,
          estimatedRefund: 80000,
          fetchTime: 607,
          aggregator: 'TEST_AGG_1',
          aggType: 'AGG',
          slippage: 2,
          sourceTokenInfo: {
            address: '0x6b175474e89094c44da98b954eedeac495271d0f',
            symbol: 'DAI',
            decimals: 18,
            iconUrl: 'https://foo.bar/logo.png',
          },
          destinationTokenInfo: {
            address: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
            symbol: 'USDC',
            decimals: 18,
          },
          fee: 1,
        },

        TEST_AGG_BEST: {
          trade: {
            from: '0xe18035bf8712672935fdb4e5e431b1a0183d2dfc',
            value: '0x0',
            gas: '0x61a80',
            to: '0x881D40237659C251811CEC9c364ef91dC08D300C',
          },
          sourceAmount: '10000000000000000000',
          destinationAmount: '25000000000000000000', // 25e18
          error: null,
          sourceToken: '0x6b175474e89094c44da98b954eedeac495271d0f',
          destinationToken: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
          approvalNeeded: null,
          maxGas: 1100000,
          averageGas: 411000,
          estimatedRefund: 343090,
          fetchTime: 1003,
          aggregator: 'TEST_AGG_BEST',
          aggType: 'AGG',
          slippage: 2,
          sourceTokenInfo: {
            address: '0x6b175474e89094c44da98b954eedeac495271d0f',
            symbol: 'DAI',
            decimals: 18,
            iconUrl: 'https://foo.bar/logo.png',
          },
          destinationTokenInfo: {
            address: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
            symbol: 'USDC',
            decimals: 18,
          },
          fee: 1,
        },
        TEST_AGG_2: {
          trade: {
            from: '0xe18035bf8712672935fdb4e5e431b1a0183d2dfc',
            value: '0x0',
            gas: '0x61a80',
            to: '0x881D40237659C251811CEC9c364ef91dC08D300C',
          },
          sourceAmount: '10000000000000000000',
          destinationAmount: '22000000000000000000', // 22e18
          error: null,
          sourceToken: '0x6b175474e89094c44da98b954eedeac495271d0f',
          destinationToken: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
          approvalNeeded: null,
          maxGas: 368000,
          averageGas: 197000,
          estimatedRefund: 18205,
          fetchTime: 1354,
          aggregator: 'TEST_AGG_2',
          aggType: 'AGG',
          isBestQuote: true,
          slippage: 2,
          sourceTokenInfo: {
            address: '0x6b175474e89094c44da98b954eedeac495271d0f',
            symbol: 'DAI',
            decimals: 18,
            iconUrl: 'https://foo.bar/logo.png',
          },
          destinationTokenInfo: {
            address: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
            symbol: 'USDC',
            decimals: 18,
          },
          fee: 1,
        },
      },
      fetchParams: {
        metaData: {
          sourceTokenInfo: {
            symbol: 'ETH',
          },
          destinationTokenInfo: {
            symbol: 'USDC',
          },
        },
      },
      tradeTxId: null,
      approveTxId: null,
      quotesLastFetched: 1519211809934,
      swapsQuoteRefreshTime: 60000,
      swapsQuotePrefetchingRefreshTime: 60000,
      swapsStxBatchStatusRefreshTime: 5000,
      swapsStxGetTransactionsRefreshTime: 5000,
      swapsStxMaxFeeMultiplier: 1.5,
      swapsStxStatusDeadline: 150000,
      customMaxGas: '',
      customGasPrice: null,
      selectedAggId: 'TEST_AGG_2',
      customApproveTxData: '',
      errorKey: '',
      topAggId: 'TEST_AGG_BEST',
      routeState: '',
      swapsFeatureIsLive: false,
    },
  },
});

export default {
  title: 'Pages/Swaps/PrepareSwapPage/ReviewQuote',
  component: ReviewQuote,
  decorators: [(story) => <Provider store={store}>{story()}</Provider>],
  argTypes: {
    setReceiveToAmount: {
      action: 'setReceiveToAmount',
    },
  },
};

export const DefaultStory = (args) => <ReviewQuote {...args} />;

DefaultStory.storyName = 'Default';
```

- checked on local ✅ 

Screenshot 2023-08-01 at 1 20 26 PM

**Review by georgewrmarshall — DISMISSED** [review]
LGTM!

**Comment by infiniteflower** [comment]
> Looks good! One suggestion to remove unused CSS. To me this approach minimizes code while aligning with the design system but you're absolutely right it's hard to work on this in isolation without trawling through the app to find the specific component and state. I tried to get the review-quote.stories.js up and running. I think doing that would provide that ability but I'm not as familiar with swaps here is where I got up to. I don't think it's necessary for this PR but it might be something worth looking into to improve the developer experience of engineers working on swaps in future

Yea that sounds like a great idea, thanks for digging into what mocks would be required for it.

**Review by dan437 — DISMISSED** [review]

**Comment by infiniteflower** [comment]
Had to rebase which dismissed all the approvals

**Review by georgewrmarshall — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]

**Pull request merged**
